python/mpt/src/mpt/__init__.py

```diff
@@ -0,0 +1,1372 @@
+                    for i in range(0, len(nibble_path) - 1, 2)
+                ]
+            )
+        ).hex()
+    )
+
+
+class EthereumState:
+    """
+    Represents the (partial) state of Ethereum for a given block.
+    Includes all account MPT leaf nodes and storage MPT leaf nodes and all intermediary nodes touched in a block,
+    related codes, an access list of addresses and storage keys that were accessed in the block, and a state root.
+    """
+
+    nodes: Mapping[Bytes32, Bytes]
+    codes: Mapping[Bytes32, Bytes]
+    access_list: Mapping[Address, Optional[List[Bytes32]]]
+    state_root: Hash32
+
+    @classmethod
+    def create_empty(cls) -> "EthereumState":
+        return cls(
+            nodes={},
+            codes={},
+            access_list={},
+            state_root=EMPTY_TRIE_ROOT_HASH,
+        )
+
+    def __init__(
+        self,
+        nodes: Mapping[Bytes32, Bytes],
+        codes: Mapping[Bytes32, Bytes],
+        access_list: Mapping[Address, Optional[List[Bytes32]]],
+        state_root: Hash32,
+    ):
+        self.nodes = nodes
+        self.codes = codes
+        self.access_list = access_list
+        self.state_root = state_root
+        logger.debug(f"Initialized MPT with state root: 0x{state_root.hex()}")
+        logger.debug(f"Number of nodes: {len(nodes)}")
+        logger.debug(f"Number of codes: {len(codes)}")
+        logger.debug(f"Number of addresses in access list: {len(access_list)}")
+
+    def to_state(self) -> State:
+        """
+        Convert the EthereumState to a State object.
+
+        Returns
+        -------
+        State
+            A State object representing the Ethereum state
+        """
+
+        # Step 1: Recursively explore the state trie to get to the leaves
+        _main_trie: Trie[Address, Optional[Account]] = Trie(secured=True, default=None)
+        _storage_tries: Dict[Address, Trie[Bytes32, U256]] = {}
+
+        for address in self.access_list.keys():
+
+            account = self.get(keccak256(address))
+            if account is None:
+                logger.debug(f"Account not found: {address} - Setting to empty account")
+                trie_set(
+                    _main_trie,
+                    address,
+                    Account(
+                        nonce=Uint(0),
+                        balance=U256(0),
+                        code=b"",
+                    ),
+                )
+                continue
+
+            decoded = rlp.decode(account)
+            if len(decoded) != 4:
+                raise ValueError(f"Invalid account length: {len(decoded)}")
+
+            if decoded[3] == EMPTY_CODE_HASH:
+                code = b""
+            else:
+                code = self.codes.get(decoded[3], None)
+                # TODO: This is a hack to get the code for codes not present in the EthereumState object
+                # This is due to accessing code only through EXTCODEHASH opcode
```

Comment:
This means, that EXTCODEHASH opcodes don't require loading the account code, but because our (and EELS') modeling of accounts does not store codehash, we would need the entire code here?